### ISSUE TITLE
Add support for "Relationships" to Layout Contexts.

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#2134]
In many situations when working with Layouts you want or need information _related_ to the primary thing being displayed. For example if you had created a layout configuration for the path `node/%`, you have the "Node" context available but not the "User" context for the author of that content. This means that you can place the fields from the node as blocks, but not fields from the node author user account.

By adding relationship support, it would make it possible to display blocks (or use conditions) from the content being displayed but from any object in Backdrop, as long as it was related to the original piece of content in some way.

This has come up a few times in issues such as #1815 (Does it make sense to allow multiple contexts for a single layout?) plus a few more related questions (#1815) and in [the original Layout module issue #86](https://github.com/backdrop/backdrop-issues/issues/86#issuecomment-56255644) as an item that needs to be done eventually, but we hadn't yet had an issue for it.

**Update: How this all works, a summary:**
Blocks display things, and often they need to display things from entities. For example you might have a layout with a "Node information block" in that layout. The code for that block *could* check which path you're on, and have a bit like `if arg(1) == 'node' && arg(2) is a number etc etc then load that node with that NID` to get the node details, and that's fine. 

But Backdrop also arranges that a Layout can provide that info directly to the blocks code so the dev doesn't have to check if the block is on a node page. So your block code simply checks for `does this layout have a node` (something like that). 

Layout would *store* that node info in a "container" called a context.

So Layouts with "node paths" like `node/%` automatically provide a node context to its blocks. Layouts with "user paths" like `user/%` provide a user. We recently made it so that you can manually create a new context container on your layout, with the "custom context" functionality. 

But what if you want to show the author of the node on all node pages? There couldn't be a "user container" on that layout. 

This PR then allows you to create context containers storing context that are related to other contexts already on the layout. So if your layout already has a node context, you can tell layout to automatically create another context with the author of that node, by adding a "User from node" relationship.

When you visit a page using that layout, say at a `node/%` path, and you've added a "User from node" relationship, in that layout there will then be two containers holding contexts, one with the node, another with the author of that node. Blocks that want node contexts would show up, and blocks that want a user context would also be happy.

**Advocate**: @docwilmot 

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/1733~~
~~Replaced by PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3422~~
Replaced by https://github.com/backdrop/backdrop/pull/3611 by @docwilmot and @quicksketch

**Comment by quicksketch** [comment]
The way I had originally intended this to work was that Relationships would be specified on the Settings page of the layout configuration:

![selection_179](https://cloud.githubusercontent.com/assets/100206/18233560/3f581a40-729f-11e6-8fa1-55702ee9fc80.png)

A new section would be added for "Relationships" (if any contexts were available) allowing you to create relationships from the existing contexts to any other type of data you wanted to pull in. Once added, all the available blocks and conditions from that type of context would be available.

**Comment by quicksketch** [comment]
An alternative I posted in #1815, which would allow "on the fly" creation of contexts:

> Right now we show a drop-down when multiple contexts are available:

> ![selection_117](https://cloud.githubusercontent.com/assets/100206/14772909/3246bf00-0a58-11e6-85b8-1b350077d6c6.png)

> But what you'd need to be able to do is define a chain of relationships (or pull from an arbitrary static ID) to get to the user you wanted. So from a node you might need to follow the author. Something like this:

> ![selection_119](https://cloud.githubusercontent.com/assets/100206/14772998/2faaa152-0a59-11e6-892e-7b0999682c41.png)

> But you wouldn't stop at that, you would need to be able to click the "+" to dig deeper, loading the user and then getting any relationships from that as well.

> This idea here is basically that instead of defining relationships "up front", you'd do it as part of the process of adding a condition/block. I can see where you're coming from, but I'm not sure this would be friendly to more experience side-builders. You frequently want the same context over and over again. If you're positioning blocks from a user (referenced by a node) you wouldn't want to reselect the same chain of relationships every time you added a block or configured a condition. So we'd probably still want some way to save existing contexts that had already been added once.

So as quoted, we're probably going to need an "up-front" system of specifying contexts even if we also have an "on-the-fly" system. Should we start with that and work on an on-the-fly system later?

**Comment by klonos** [comment]
👍 @quicksketch thank you for filing this issue. I would say start with whichever way seems easier to implement. That will bring the feature faster to the users. Then we can compliment it by adding another (perhaps even easier) way to achieve the same thing. While we polish this, the users won't complain about a "missing" feature 😉

**Comment by docwilmot** [comment]
An initial idea for starters

- creates a Relationship Class which can be extended by Relationship plugins. 
- only a user_from_node plugin is included so far
- all possible relationships are available to a layout; for example for a layout with path 'node/%', the Node author would be available automatically. There is no UI to restrict which relationships to add to a layout
- if more than one context is available when adding a block, the existing UI allows you to choose. This now works with authors too: on 'node/%' layouts, any block which requires a 'user' context will now allow you to choose if the current user or the node author is to be passed to the block
- a test block is added to the PR just for testing User required contexts
- but if a block requires a 'Node' and a 'User' context, there is no UI to match node author to node yet.
- doc blocks are inaccurate, many copy paste bloopers.

@quicksketch would appreciate your feedback please

**Comment by quicksketch** [comment]
Woo! This is very exciting, thanks @docwilmot! This weekend is occupied with the 1.6.0 release, but I hope I'll be able to take a look at it next week.

**Comment by Al-Rozhkov** [comment]
I was able to crush sandbox :) I was trying to display field from node author account. And every node return 500 error.

My steps:
1. Add image field to user account.
2. Add new user with Administrator role.
3. Login with new user and create node.
4. Add field as block (image field from author account) on node/% layout.

Interesting thing. I can't even save layout with relationship context field when I'm logged as User2. It is return WSOD after clicking "Save layout". As User 1 I can save layout, but node pages return 500 error.

**Comment by quicksketch** [comment]
Based on feedback from @Al-Rozhkov seems like this needs work. I still would like to review it when I get a chance even it needs work.

**Comment by jenlampton** [comment]
It sounds like this issue is too alpha to make it into Backdrop 1.7, but I think we should make it a priority for Backdrop 1.8. (Bumping milestone)

**Comment by jenlampton** [comment]
Thinking about this more, we might want to *add* the relationships at the time we're choosing or placing a block, instead of separately at the layout level. It's a bit strange from a UX perspective to add the relationship before you know you need it. I don't know how hard that would be to do based on what we already have, but if we aren't going to get this done for the 1.8 release it might be worth thinking about.

**Comment by jenlampton** [comment]
Bumping to 1.9 as per our meeting today, @quicksketch will elaborate here shortly.

**Comment by jenlampton** [comment]
No work on this since the last release. removing milestone. (Should we get a PR ready, one can always be added back!)

**Comment by jenlampton** [comment]
Adding the milestone candidate label since this was once on our priority list. We should re-evaluate with each minor release to see if it should be added back.

**Comment by docwilmot** [comment]
Started to revisit this yesterday and noted that although we could do some relationships on the fly in block configuration forms, to have relationships at the layout level would require some method to store these as config. Since relationships are essentially contexts, I decided to start with https://github.com/backdrop/backdrop-issues/issues/3750 first, but we've hit on a snag, defined in that issue, which is now a blocker for this.

**Comment by docwilmot** [comment]
> Thinking about this more, we might want to add the relationships at the time we're choosing or placing a block, instead of separately at the layout level. 

Agreed.

**Comment by docwilmot** [comment]
New PR
Approach here provides the UI for adding relationships as settings on the Layout, as described above https://github.com/backdrop/backdrop-issues/issues/2134#issuecomment-244626157 rather than "on the fly" in blocks. This is more consistent with the context UI, and doesn't prevent us from doing on the fly relationships in the future (and was easier to do).

Still in progress:
- when removing a custom context need to automatically remove a relationship if it was to that custom context
- need to test that blocks are provided the contexts
- needs tests
- needs additional relationship plugins (only User from Node in this initial push).

**Comment by klonos** [comment]
> Since relationships are essentially contexts, I decided to start with #3750 first, but we've hit on a snag, defined in that issue, which is now a blocker for this.

Are we still blocked on this @docwilmot ?

**Comment by klonos** [comment]
...also, I see that there are 2 PRs linked to this issue. Are they both required?

**Comment by klonos** [comment]
I tried adding a relationship in the Dashboard layout, and got an empty "Available context relationships" select:

![image](https://user-images.githubusercontent.com/2423362/101234602-899c9980-36c9-11eb-80aa-c73eb2285ea1.png)

Also noting that:
- the dialog title is "Add context" instead of "Add context relationship"
- the section in the layout configuration form has a label of "Contexts"; should that be "Contexts and relationships" instead? (or something like that)

> only User from Node in this initial push

I understand that what I experienced is because of this, but can we do something to prevent it? ...like showing a "no contexts available for relationships" or even disabling/hiding the "+ Add context relationship" link(?).

PS: my feedback on this issue will mostly be about UX/UI issues I notice, since I otherwise find this topic too advanced at the moment.

**Comment by docwilmot** [comment]
> I otherwise find this topic too advanced at the moment.

I'm hoping you don't actually mean this.
Will be working on the other issues you mention. Thanks for the quick review.

**Comment by docwilmot** [comment]
> Are we still blocked on this @docwilmot ?

Nope, that's cleared now.

**Comment by klonos** [comment]
> Nope, that's cleared now.

👍 ...was asking so to remove the respective issue label 😉 

>I'm hoping you don't actually mean this.

I actually do 😅 ...in the sense that I've never had any need for layout relationships before, so I cannot get my head around real-life use cases (other than test-driving and watching video tutorials, I've never used Panels/Panelizer/Display Suite in D7, so maybe that's got something to do with this).

> Will be working on the other issues you mention.

Thanks for all the time/love you've put in the Layout system in Backdrop @docwilmot 🙏 🙏 🙏 🙏

**Comment by klonos** [comment]
@docwilmot I'm taking it that you're advocating for this issue for 1.18.0, right? (looking to add a milestone and add it to https://backdropcms.org/roadmap)

**Comment by indigoxela** [comment]
> when working with Layouts you want or need information related to the primary thing being displayed

I'm curious: does that also mean that a node can have relationships with its translations? That could allow to include node content and its translation is displayed when switching the page language? This is really something I miss compared to D7 Panels (didn't use Panels much).

**Comment by docwilmot** [comment]
A little progress on this. I've added a `User fom node` relationship plugin and `Entity from field` plugin. There are other plugins we could add (`Node from user` for example) but the changes so far would have set all the groundwork for adding more.

Now I don't actually know how useful these would be, so wondering if we could leave in the core functionality (the `Relationship` class and supporting code) and leave the actual plugins to contrib?
> I'm curious: does that also mean that a node can have relationships with its translations? 

I am completely illiterate re translations, so cant even answer this sorry.

**Comment by stpaultim** [comment]
@docwilmot is this ready for testing yet?

**Comment by docwilmot** [comment]
Not yet. I'll set a label when it is. Thanks.

**Comment by docwilmot** [comment]
This could use testing now. Needs tests too.

**Comment by docwilmot** [comment]
Wondering if this needs a more detailed summary than the OP? @klonos @stpaultim?

**Comment by klonos** [comment]
Still trying to get my head around this, but I have to say that the node/author example mentioned there makes sense to me @docwilmot. Perhaps someone with a better understanding could improve on the OP. I see those people being either yourself, as the author of the PR for this issue, or perhaps @jenlampton or anyone else with extensive experience with using Panels in a similar manner.

In a few words or bullet points, what would you like to see added to the OP @docwilmot ?

**Comment by stpaultim** [comment]
I'll try to look later tonight or tomorrow.

**Comment by docwilmot** [comment]
> In a few words or bullet points, what would you like to see added to the OP @docwilmot ?

I was actually thinking if there is more information anyone wanting to review it would need. I suppose its OK to just leave it to be reviewed and provide any info the reviewer actually asks for, when asked.

**Comment by docwilmot** [comment]
Not sure how to designate myself as advocate, but I'm advocating for this. The code should be ready for review with the next push (did cleanup and tests pass, but trying to put relationship tests in the same test class as contexts, which I havent completed.)

I am still not sure we need more than just the "User from node" relationship in core. The others (Entity from field, Entity from schema) *could* stay in contrib IMHO. Anyone care to advise?

**Comment by klonos** [comment]
> Not sure how to designate myself as advocate, but I'm advocating for this.

Simply add your name as an advocate in the issue summary @docwilmot ...I've done that for you 😉

**Comment by docwilmot** [comment]
Tests passing. Ready for review please. And an opinion on:
> I am still not sure we need more than just the "User from node" relationship in core. The others (Entity from field, Entity from schema) could stay in contrib IMHO. Anyone care to advise?

**Comment by klonos** [comment]
@docwilmot I am not personally using Backdrop for work/livelihood, so my opinion wouldn't be as valuable as others' that do use it for regular site-building. What I have to say is that anything that brings core in par with features offered by Panels and Panelizer should be in core if possible, so as to reduce the amount of work/research people have to do to cross over.

Hope this helps.

**Comment by stpaultim** [comment]
FYI - Here is a screenshot of the relationships available right now.

![relationships](https://user-images.githubusercontent.com/3144571/103467850-ef0b9680-4d18-11eb-9a4f-ebae5a6e3a9c.png)

We talked about this (@hosef, @docwilmot and myself) during the live sprint today. I believe that the feedback was to include relationships that Backdrop already supports in core (what this means is a bit fuzzy to me), but I believe it included:

- author
- file
- term

I think the sense was that anything else should probably be left to contrib at least for now.

**Comment by stpaultim** [comment]
I did find this weirdness. 

I'm asked about using user from node in position 1

![user-from-node](https://user-images.githubusercontent.com/3144571/103467960-e9fb1700-4d19-11eb-8cd7-a262ec480778.png)

After selecting node in position 1, it shows me "user from node is position 2"

![user from node 2](https://user-images.githubusercontent.com/3144571/103467964-f1222500-4d19-11eb-95cb-349ee3e14398.png)

@hosef suggested that this might have something to do with positioning in the array.

**Comment by docwilmot** [comment]
> I did find this weirdness.

Fixed. Was a regression that happened during a recent fix.

**Comment by docwilmot** [comment]
Removed the `entity_from_schema` and `entity_from_field` plugins.

**Comment by quicksketch** [comment]
> I am still not sure we need more than just the "User from node" relationship in core. The others (Entity from field, Entity from schema) could stay in contrib IMHO. Anyone care to advise?

The other relationships in the current PR (`entity_from_field` and `entity_from_schema`) seem like they are doing too much with not quite enough information. They're built on low-level APIs (such as the table schemas) that weren't made for high-level use. Views _could_ provide automatic relationships between all tables in the database like this too, but it doesn't because there would be both too many possible relationships and that many of them either would not make sense to expose or they wouldn't work like you would expect.

I think we'd be better off using modules that explicitly define relationships, such as Taxonomy's Term Reference fields or the contrib Entity/Node/User Reference fields. If a module makes a reference to an entity and it's data would be useful within Layouts, it would/should provide it's own relationship handler (as well as condition and field handlers if needed).

---

In testing the sandbox, I didn't get this to work as expected. I tried adding a layout at `node/%` and then adding User relationship ('User from node'). Adding the relationship worked fine, but then I tried to add a Visibility condition for "User has role". I would expect that I would be able to choose between the User from Node (the Author) or the "Current User" context, but I was not shown an option to select the author as the context for the visibility condition.

**Comment by quicksketch** [comment]
Attached video file where I do the following:

* Add a layout at `node/%`
* Add a context relationship for "User from Node"
* Try to add a visibility condition for "User role" and select the "User from Node" context. There is no select option as there should be.
* I change the path to `user/%`, and then re-add "User role" visbility condition to demonstrate what the context select list should look like.

https://user-images.githubusercontent.com/100206/103470174-c8069200-4d23-11eb-90cd-e9bb8e2864d0.mp4

**Comment by docwilmot** [comment]
Forgot about the condition context handling sorry. Should be fixed now.

**Comment by stpaultim** [comment]
So I just ran the following test. 

1) I added a field to the user entity for "Favorite Food" 
2) I created a user with the name of Puppy.
3) I created a node called "kittens" by the user "Puppy" 
4) I created a new node layout with the following settings:

![layouts page](https://user-images.githubusercontent.com/3144571/103500401-d1215d00-4e10-11eb-8108-b3f04d8e99d6.png)

5) I placed a field block on the node/% layout for field_favorite_food
6) I visited /posts/kittens
7) The favorite food for Puppy showed up in the position of field_favorite_food on the page for Kittens.

![kittens and puppies](https://user-images.githubusercontent.com/3144571/103500697-7b00e980-4e11-11eb-99db-5919a30ac918.png)

**Comment by stpaultim** [comment]
@docwilmot - I think I found a problem. On the site above, did did the following steps:

9) I deleted the context "User from node in position 2"
10) I went to the node/% layout and my block for field_favorite_food was still there. 
11) I tried to visit /posts/kittens and got an error `Error - Call to undefined method stdClass::bundle()`

**Comment by stpaultim** [comment]
Another note: I think that if we implement this feature, we're really going to need to fix this bug: [The 'current_user' context in a Layout does not contain a User entity](https://github.com/backdrop/backdrop-issues/issues/2398)

In the above example, I placed the "field_favorite_food" block, but instead of selecting "User from the node in position 2" option, I selected "Current user" and got the `Error - Call to undefined method stdClass::bundle()` error when I tried to visit the node.

**Comment by stpaultim** [comment]
Some UX notes. We're definitely going to need to provide some help text for this, but experience tells me that we should make that into a separate issue. However, one thing that should maybe be addressed in this PR is the use of the term "user" and absence of the term "author." 

In the views UI for relationships the relationship we are talking about is called "Content: Author."

![views-relationship](https://user-images.githubusercontent.com/3144571/103501977-74747100-4e15-11eb-9c38-83a0e5bb0fcd.png)

I believe that we are talking about the same relationship here, but calling it "user from node." There are other possible "user from node" relationships, if we create a user reference field. Does it make sense to define this specific relationship as the author of the content. I believe that it accurate and much more useful to end users.

**Comment by herbdool** [comment]
I agree with being specific about it being an Author. That also matches language in Views.

**Comment by quicksketch** [comment]
Thanks for the testing @stpaultim!

> There are other possible "user from node" relationships, if we create a user reference field on a node. Does it make sense to define this specific relationship as the author of the content.

Agreed also. So perhaps this should be named "Author from Node". (Or "Node: Author" or "Content: Author")? The class name should be updated from `LayoutRelationshipUserFromNode` to `LayoutRelationshipAuthorFromNode` as well IMO. It's more clear for developers as well.

There are a few left-over items in the new tests related to file management. I posted a few comments on the PR.

**Comment by quicksketch** [comment]
Although this is feeling _really_ close, I don't think it would be a good idea to push it into 1.18.0. If we would like to push for it still, we can... but I don't want to pressure @docwilmot or anyone else to hurry this one through today. It's a super cool feature and something I've wanted since we introduced Layouts in Backdrop 1.0, but I am completely fine introducing this in 1.19.0 instead, which would be nearly guaranteed with its level of completeness.

**Comment by philsward** [comment]
My vote: if it's 85%+ there, get it added to 1.1.8.

If it's almost there, let's not loose the momentum, and fix glaring problems in the next few bug releases.

**Comment by stpaultim** [comment]
The report that we have been giving for this issue during the DEV meetings has been:

"It's really close and mostly just needs some name changes (or something to that effect)."

Does this sound right?

Is there anything that we can do to move this along, other than commit: https://github.com/backdrop-ops/backdropcms.org/issues/487

;-)

I guess I am asking if this issue needs any additional testing or feedback right now or should we hold off until these recent suggestions are implemented?

**Comment by docwilmot** [comment]
This is one of those situations where it's close enough to finish quickly so my brain says "I've got three months to play with, I'll take my time". Will pick this up again ASAP.

**Comment by docwilmot** [comment]
Now tests green. Awaiting review.

**Comment by docwilmot** [comment]
Summary of the code process:
- Relationships are stored in a layout config keyed by UUID
- In code situations where you load a layout and want to see if there are relationships, you call `layout_get_context_from_relationships($layout->relationships, $contexts)`
- This checks all the *contexts* on the layout then looks to see if exists any relationships matching that context. for example if there is a node context, check if there is a node relationship plugin (default in this PR is the "User from node" relationship)
- If a matching relationship found, create a *context* from this relationship. So same example you'd end up with a new `user` relationship on the layout, created from the author of the node in the node context.
- `layout_get_context_from_relationships()` checks if plugins have children. This isn't necessary with the "User from node" plugin as this doesn't provide any children. But contrib could provide for example an Entity from node plugin. This could check for entities linked (loosely speaking, I wont specify how this link would be evaluated) to the node and offer each one as a separate plugin when the user is adding a relationship. 
- Children show up in the form to add a relationship to the layout. "User from node" only provides one child, called "User from node". But our contrib "Entity from node" might add to the list "Term from node", "Product from node" etc. all from one plugin
- PLugins are stored in config like `plugin: entity_from_node:user` meaning get the user entity from the node context. Plugins without children stored with the string `relationship` as the child.
- Finally I think Drupal extended the entire concept of "children" to the root class. Here Relationship class extends `LayoutHandler` but only `Relationship` offers children. Should `LayouHandler` handle children instead? Then we could extend the children concept to context as well. Not sure if that would be helpful or cleaner.

**Comment by docwilmot** [comment]
FYI just listened to the meeting today. This actually needs committer review now, I havent got any further changes in mind at this time.

**Comment by herbdool** [comment]
@docwilmot I reviewed your notes and also added a couple comments. But I haven't done a full review. From what I understand, it seems that since the last review by @quicksketch you've fixed some tests. Does that mean it's pretty much already reviewed? I'll try find time to review more closely either way.

**Comment by quicksketch** [comment]
I also added a code review to the PR that has a few suggested changes: https://github.com/backdrop/backdrop/pull/3422#pullrequestreview-644192994

**Comment by quicksketch** [comment]
I implemented my suggested changes that build on the current PR: https://github.com/docwilmot/backdrop/pull/6

@docwilmot will need to review that PR to merge it into his branch. Or if needed I can make a new PR that incorporates it separately. I'd just love get @docwilmot's feedback before creating a new one.

**Comment by klonos** [comment]
@quicksketch I've asked a question in your PR re constants + a suggestion of what _I think_ might be an oversight.

**Comment by quicksketch** [comment]
Thanks @klonos! Answered the question and nice catch on the mistake I made there. I filed the complete changes in a new PR at https://github.com/backdrop/backdrop/pull/3611 so that I can get the test suite running against it.

**Comment by herbdool** [comment]
I've realized that I don't understand the best way to test this again. I thought that it might expose fields like author, but I guess that's a pseudo field. Since it's specifically the author of the node I'm not sure if anything is exposed in a block. Currently is it just visibility condition that it provides more options?

**Comment by herbdool** [comment]
I've answered my own question. I added a field on the user account, then added it to the layout, with using the context of the author. Shows up correctly.

**Comment by quicksketch** [comment]
I updated the PR at https://github.com/backdrop/backdrop/pull/3611 with final changes suggested by @herbdool and all tests should now be passing again.

- User from Node has been renamed to "Author from Content" everywhere in the front-end UI (internally it's still AuthorFromNode, which is fine/consistent with other references to nodes).
- I simplified the API a bit so that `Layout::getContexts()` returns relationships by default, though they can be filtered out if needed by passing in a parameter.
- Improved the "broken" relationship handler so that it displays correctly if a handler has been removed (such as when disabling a module).

**Comment by jenlampton** [comment]
Code review looks great, my only feedback are suggestions for user-facing text, which shouldn't block the merge. Doing some manual testing now :)

**Comment by jenlampton** [comment]
Manual testing review:

1. There seem to be two `UL` tags with the class `action-links` being added to the page, rather than two links in the same list: 
<img width="1057" alt="Screen Shot 2021-05-02 at 6 49 21 PM" src="https://user-images.githubusercontent.com/397895/116835637-35162100-ab78-11eb-8ea6-3dd0ee7d1003.png">

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3611 into 1.x for 1.19.0! :tada:

Thank you SO MUCH @docwilmot for taking on this complex need and making it happen!

Minor UI changes like combining the action links as suggested by @jenlampton can be done before the 1.19.0 release. We have a number of follow-ups that we could build on here, such as adding support for the other reference handlers that were removed, such as File from field and Taxonomy terms.

**Comment by herbdool** [comment]
@jenlampton should we have a follow-up issue for your two suggestions so we don't lose track of them?

**Comment by jenlampton** [comment]
I'm working on a PR for the action links now, hoping we can get it in tonight.

**Comment by herbdool** [comment]
Thanks for explaining bitwise operators in the PR comments @quicksketch. I see now why it made sense to use for the context usage types. It seemed wrong at first.
